misc/quick-protobuf-codec/src/lib.rs

```diff
@@ -35,16 +39,59 @@ impl<In: MessageWrite, Out> Encoder for Codec<In, Out> {
     type Error = Error;
 
     fn encode(&mut self, item: Self::Item<'_>, dst: &mut BytesMut) -> Result<(), Self::Error> {
-        let mut encoded_msg = Vec::new();
-        let mut writer = Writer::new(&mut encoded_msg);
-        item.write_message(&mut writer)
-            .expect("Encoding to succeed");
-        self.uvi.encode(Bytes::from(encoded_msg), dst)?;
+        write_length(&item, dst);
+        write_message(&item, dst)?;
 
         Ok(())
     }
 }
 
+/// Write the message's length (i.e. `size`) to `dst` as a variable-length integer.
+fn write_length(message: &impl MessageWrite, dst: &mut BytesMut) {
+    let message_length = message.get_size();
+
+    let mut uvi_buf = unsigned_varint::encode::usize_buffer();
+    let encoded_length = unsigned_varint::encode::usize(message_length, &mut uvi_buf);
+
+    dst.extend_from_slice(encoded_length);
+}
+
+/// Write the message itself to `dst`.
+fn write_message(item: &impl MessageWrite, dst: &mut BytesMut) -> io::Result<()> {
```

Comment:
Why is this custom code needed? Why can one not use [`serialize_into_slice`](https://docs.rs/quick-protobuf/latest/quick_protobuf/writer/fn.serialize_into_slice.html)?

Reply:
`serialize_into_slice` performs its own length-prefixing :(

Reply:
Thanks! That is a bummer.

For others, this is further discussed in https://github.com/tafia/quick-protobuf/issues/202.

I am hesitant to add this functionality to `quick-protobuf-codec`. I think it belongs into `quick-protobuf`. I don't think we should duplicate this logic here, especially since it requires `unsafe` code.

@thomaseizinger what do you think is the probability that the maintainer of `quick-protobuf` accepts this as a contribution? In other words do you think it is worth pushing https://github.com/tafia/quick-protobuf/issues/202 forward?

Reply:
I think there are several improvements to be made / upstreamed, see https://github.com/tafia/quick-protobuf/issues/246.

I don't think it is worth blocking these optimisations on coming to an agreement and upstreaming a better API.

Especially the bit about writing to a slice of `MaybeUninit<u8>` will probably attract some bike-shedding.

I'd rather land these first, reap the benefits and then make the code nicer by upstreaming them. It also gives us something to point at and say "that is the API we'd like to have in `quick-protobuf-codec`".